File: haskell/glue/src/Glue/Lib/Math/Trigonometric/Atan.hs
module Glue.Lib.Math.Trigonometric.Atan where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Arctangent function
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Atan.atan exactly
atan :: IR Eval
atan = NativeFunc atanImpl

-- Arctangent function implementation
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Atan.atanImpl exactly
atanImpl :: IR Eval -> Eval (IR Eval)
atanImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.atan (fromIntegral n))
    Float n -> pure $ Float (Prelude.atan n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/trigonometric/atan.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Arctangent function (returns radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Atan.atan exactly
final Ir atan = IrNativeFunc(atanImpl);

/// Arctangent function implementation (returns radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Atan.atanImpl exactly
Eval<Ir> atanImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrFloat(math.atan(n.toDouble()))),
    IrFloat(value: final n) => Eval.pure(IrFloat(math.atan(n))),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
